add redis channel

JunLuo opened the pull request:
In some cases, we need events can not be lost. But we don't want to install zk & kafka. 
To make event deliver only once and to deploy many flume agents in different nodes but share only one channel, we developed redis channel. 
We already use the redis channel in out production environment, and we want to feed back to flume.

JunLuo committed:
add redis channel

add redis channel

bessbd commented:
Hi @JunLuo ,

Thanks for the pull request!

This looks like a very useful change to me. However, Travis CI has reported a compilation error:

> [ERROR] COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> [ERROR] /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[3,22] error: package junit.framework does not exist
> [ERROR] /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[4,22] error: package junit.framework does not exist
> [ERROR] /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[5,22] error: package junit.framework does not exist
> [ERROR] /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[11,12] error: cannot find symbol
> [ERROR]  class TestCase
> /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[26,18] error: cannot find symbol
> [ERROR]  class AppTest
> /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[28,19] error: cannot find symbol
> [ERROR]  class AppTest
> /home/travis/build/apache/flume/flume-ng-channels/flume-redis-channel/src/test/java/org/flume/redis/channel/AppTest.java:[36,8] error: cannot find symbol
> [INFO] 7 errors 

Can you please fix it, so that we can proceed with the reviews?


Thank you,

Donat

JunLuo committed:
delete unused testapp

delete unused testapp

JunLuo committed:
add license comments

add license comments

JunLuo committed:
fix codestyle check errors

fix codestyle check errors

JunLuo committed:
fix codestyle check errors

fix codestyle check errors

JunLuo committed:
fix codestyle check errors

fix codestyle check errors

JunLuo committed:
init redis channel version

init redis channel version

JunLuo committed:
fix a stupid error

fix a stupid error

JunLuo commented:
Hi @bessbd ,
I have fixed the errors about Junit and code style :)

JunLuo committed:
fix a bug

fix a bug

bessbd commented:
Hi @JunLuo ,

Thank you for the changes!

On a first pass, I've noticed that a new user guide entry is missing.
For this change to get merged, some documentation would be necessary.
For an example, I'd recommend [Kafka Channel](http://flume.apache.org/FlumeUserGuide.html#kafka-channel).
Do you think you can make the necessary additions?


Thank you,

Donat

JunLuo committed:
add redis channel documentation

add redis channel documentation

JunLuo commented:
Hi @bessbd ,
I've add some documentation about redis channel :)

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+ */
+package org.apache.flume.channels.redis;
+
+import com.google.common.base.Optional;
```
Is it possible to use java 8 optional instead?

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+    @Override
+    protected void doPut(Event event) throws InterruptedException {
+      type = TransactionType.PUT;
+      if (!serializedEvents.isPresent()) {
```
Why not just initialize with the empty list?

or we could use the Optional.ifPresent(Consumer<? super T> consumer) at the end, where the get() is called

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+      }
+
+      try {
+        if (!tempOutStream.isPresent()) {
```
the optional wrapper seems unnecessary because there is a single initialization and 3 get in the same method.
Initialization could be moved to declaration

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+        if (!tempOutStream.isPresent()) {
+          tempOutStream = Optional.of(new ByteArrayOutputStream());
+        }
+        if (!writer.isPresent()) {
```
the optional might be an overhead here too.

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+      } catch (Exception ex) {
+        LOGGER.warn("Error while get multi message", ex);
+      }
+      if (!events.isPresent()) {
```
This variable has only this initialization and a few get in this class. I guess the Optional here is not needed.

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,351 @@
+                message.getBytes());
+            decoder = DecoderFactory.get().directBinaryDecoder(in,
+                decoder);
+            if (!reader.isPresent()) {
```
same one initialization with gets.

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/tools/RedisDao.java:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  public static synchronized RedisDao getInstance(String host, int port, String passwd) {
+    if (null == instance) {
```
What happens if you would like to have two redis channel with different redis servers?

szaboferee commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -3391,6 +3391,36 @@ Also please make sure that the operating system user of the Flume processes has
+.. code-block:: properties
+
+    a1.channels.channel1.type = org.flume.channels.redis.RedisChannel
+    a1.channels.channel1.server = redis-server
```
It might be better to have an example host (localhost) here or an IP address

JunLuo replied:
fixed

szaboferee commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -3391,6 +3391,36 @@ Also please make sure that the operating system user of the Flume processes has
+
+================================================  ================================  ========================================================
+Property Name                                     Default                           Description
+================================================  ================================  ========================================================
```
The cluster scenario is not documented

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/tools/RedisController.java:
```diff
@@ -0,0 +1,92 @@
+  private int maxWaitMillis;
+
+  private JedisPoolConfig createJedisConfig() {
+    redisConnTimeout = 60000;
```
It would be great if these settings could be set from the configuration

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/tools/RedisController.java:
```diff
@@ -0,0 +1,92 @@
+    return jedisConfig;
+  }
+
+  public static RedisController getRedisController(String host, int port,
```
What happens if we would like to have two channel with different redis servers?

JunLuo replied:
fixed

szaboferee commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/tools/RedisController.java:
```diff
@@ -0,0 +1,92 @@
+    maxIdle = 300;
+    minIdle = 10;
+    maxWaitMillis = 60000;
+    JedisPoolConfig jedisConfig = new JedisPoolConfig();
```
I see the cluster related config names but not the actual implementation that uses them.

JunLuo replied:
fixed

szaboferee reviewed (commented):
Thank you for the new channel, it is going to be very useful!
I added some questions and suggestions bellow.

It would be great if you could write unit and/or integration test.
https://github.com/kstyrc/embedded-redis
Could be a good tool to create the integration tests.

JunLuo committed:
remove useness optional wrapper

remove useness optional wrapper

JunLuo committed:
do many changes

`add sentinel and cluster type
`move redis config to channel config
`add unit test

JunLuo committed:
fix code style

fix code style

JunLuo committed:
fix code style error

fix code style error

JunLuo committed:
add cluster and redis config documention

add cluster and redis config documention

JunLuo commented:
Hi @szaboferee ,
I've done these changes:
1. Add cluster and sentinel mode
2. Remove useless optional
3. Add the documention about cluster and sentinel
4. Move redis config to flume configuration
5. Fix the error of singleton
6. Add unit test

JunLuo commented:
Hi @szaboferee ，
 I've fix all problems you mentions in commits, What should I do next?

bessbd commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -3391,6 +3391,61 @@ Also please make sure that the operating system user of the Flume processes has
+Redis Channel
+~~~~~~~~~~~~~
+
+The events are stored in a redis server (must be installed separately). Redis provides high concurrency. If you have many agents but want to share one channel which can provide high concurrency, Redis Channel is a good choice. Redis also can provide high availability and replication, you can choose the best way to adapt your application.
```
Should be "Redis" instead of "redis".

bessbd replied:
Also, the sentence `Redis provides high concurrency.` seems redundant.

bessbd commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -3391,6 +3391,61 @@ Also please make sure that the operating system user of the Flume processes has
+
+.. code-block:: properties
+
+    a1.channels.channel1.type = org.flume.channels.redis.RedisChannel
```
We could add RedisChannel to `flume/flume-ng-configuration/src/main/java/org/apache/flume/conf/channel/ChannelType.java` so that people don't have to type `org.flume.channels.redis.RedisChannel`

bessbd commented on flume-ng-channels/flume-redis-channel/src/main/java/org/apache/flume/channels/redis/RedisChannel.java:
```diff
@@ -0,0 +1,462 @@
+      redisOperator = RedisInit.getInstance(redisConf, jedisPoolConfig);
+      counter.start();
+      super.start();
+    } catch (Exception e) {
```
I have noticed a few occurrences of `catch (Exception e)`. According to https://stackoverflow.com/a/2416334/5323166 we should catch the most specific type we can. Can we catch more specific exceptions here?

bessbd commented:
Overall, this change looks fine to me. After receiving feedback for the comments I have left, I'd like to request contributors and committers to do a review of this PR so that we can merge it.

JunLuo committed:
change redis description

change redis description

JunLuo committed:
change channel type

change channel type

JunLuo committed:
channel exception type

channel exception type

JunLuo commented:
Hi @bessbd ,
I've fix all problems you mentions in commits. :)

asfgit commented:
Can one of the admins verify this patch?

rgoers commented:
I will make a note to myself to take a look at this one. I would expect that if it is accepted it will end up in a logging-flume-redis repo as we are trying to move things out of the main Flume build that won't necessarily have the same build lifecycle.
